### PR TITLE
dc.molnet module

### PR DESCRIPTION
**Pull request by miaecle**
This PR adds in:

- dc.molnet module: this module includes all the data loading scripts and a new function `run_benchmark`

- RawFeaturizer: a dummy featurizer directly outputs rdkit mol struct or smiles string

Several changes:

all the dataset files are now available on aws. Calling `run_benchmark` function will auto detect and download dataset files to designated folder(need to set DEEPCHEM_DATA_DIR in your environment, otherwise will be downloaded to /tmp)

`run_benchmark` is available(not tested yet) to accept user-defined featurizer and models. For featurizer, it could either be a dc.feat.Featurizer class(from inheritance) which has the function `_featurize` that transform a single rdkit mol struct to features, example:

```
from rdkit import Chem
from deepchem.feat import Featurizer

class RawFeaturizer(Featurizer):
  def __init__(self, smiles=False):
    self.smiles = smiles
    
  def _featurize(self, mol):
    if self.smiles:
      return Chem.MolToSmiles(mol)
    else:
      return mol
```

or it can work on batch of data by setting `featurizer='Raw'` and then directly works on dataset struct. For models, it should have function `fit` and `evaluate` which can accept deepchem dataset struct.

I also added another script `benchmark2.py` based on the new module in the examples folder, which could fulfil the work of `benchmark.py`(under test now). Currently this PR is not destructive, though I tend to refine the examples folder.

### CODE REVIEW

**Comment by XericZephyr** [comment]
Hi, @miaecle, 

I personally think you can, by default, use some other folder like `~/.deepchem/datasets/` and give the user that information. (Use `os.path.expanduser("~/.deepchem/datasets/")` to get full path.)

`/tmp` can be easily purged by a reboot and hence cause frustration for users.

File: deepchem/feat/raw_featurizer.py [hunk 1]
```diff
@@ -0,0 +1,19 @@
+#!/usr/bin/env python2
+# -*- coding: utf-8 -*-
+from __future__ import print_function
+from __future__ import division
+from __future__ import unicode_literals
+
+from rdkit import Chem
+from deepchem.feat import Featurizer
+
+class RawFeaturizer(Featurizer):
```
Comment (rbharath) [inline deepchem/feat/raw_featurizer.py]: This is a great addition! Good thought putting this in :-)

File: examples/benchmark2.py [hunk 2]
```diff
@@ -0,0 +1,108 @@
+#!/usr/bin/env python3
```
Comment (rbharath) [inline examples/benchmark2.py]: This looks really nice! Users can easily make their own benchmarking schema based off this template.

**Comment by rbharath** [comment]
@XericZephyr You should just be able to set `DEEPCHEM_DATA_DIR` in your `~/.bashrc` and it won't write to `/tmp`

**Comment by rbharath** [comment]
@miaecle This looks excellent! Props for getting this through so quickly. It looks like it's failing on `yapf`. Would you mind fixing up the syntax issues? We can go ahead and merge this PR in after.

**Comment by XericZephyr** [comment]
Hmm, @rbharath, you don't want to always ask your user to set $DEEPCHEM_DATA_DIR in `~/.bashrc`. 
Also, some linux distribution does not support `~/.bashrc` startup script.

I believe changing the default out of `tmp` can potentially reduce the risk seeing future issues titled like "DeepChem constantly redownload my dataset." or "Can deepChem cache the downloaded dataset?"

It is also not engineering painful by adding the following code: 
```python
default_datapath = os.path.expanduser("~/.deepchem/datasets/")
if not os.path.exists(default_datapath):
    os.makedirs(default_datapath)
datapath = default_datapath
```

**Comment by miaecle** [comment]
@XericZephyr Thanks for your suggestion! I think we do not want to use a folder without user's designation.(dataset files could be huge) Setting up `/tmp` as our default is a reasonable alternate. Also, since downloading the dataset typically takes less than 1% of the full benchmarking time, we believe it is reasonable.

**Comment by rbharath** [comment]
@miaecle Thanks for quick turnaround! Going to go ahead and merge.

**Comment by XericZephyr** [comment]
Hi, @miaecle, thanks for the comments and explanation. I am mostly considering using the benchmark in a distributed cluster for efficiency. That's where most of my suggestions come from. 

For the above, some cluster does not allow users to access `/tmp` so we have to set `$DEEPCHEM_DATA_DIR` on such environment.

Datasets could be large. So I have to confess directly downloading the dataset to user's folder might not be a perfect idea as well.

A constructive feedback might be we need to document the `/tmp` problem and let the user know they might lost the datasets after a reboot on linux system and it is almost impossible to use `/tmp` on a Windows system. A solution is to set `$DEEPCHEM_DATA_DIR`.

**Comment by evanfeinberg** [comment]
Great job Michael!

Does this add a pdbfixer dependency? If so it should likely be added to the install section of readme.md

**Comment by XericZephyr** [comment]
GJ @miaecle as well.

Do we need to remove the `examples` folder since most of them have been moved to `molnet` package?

**Comment by rbharath** [comment]
@evanfeinberg The PDBFixer dependency comes from other changes. Definitely agree that needs to be added to README though. Will raise an issue.

@XericZephyr I like the idea of keeping around the examples folder to show users how to do this analysis themselves. It's a good resource for more advanced users (we make heavy use of it internally)

**Comment by miaecle** [comment]
@XericZephyr You are right, we need to remind users to setup dependency for deepchem if `/tmp` settings may cause any unwanted troubles. For `examples` folder, I generally agree with Bharath, also these are the simplest scripts if one want to try playing around with our package(may serve as tutorial).
@evanfeinberg Thanks! I actually directly uses the featurized datasets for molnet pdbbind loader, pdbfixer is thus not included in this part(though it is still required for the whole deepchem package)

**Pull request merged**